### ISSUE TITLE
keep integration (4 widgets) working with Drupal 8 version (under development) of OpenScholar

### ISSUE THREAD

**Issue by rbran100** [IQSS/dataverse#6381]
Hey guys, we are updating the dataverse widgets for the D8 version of our openscholar code and were hoping you could make some small changes to (https://dataverse.harvard.edu/resources/js/widgets.js).

See: https://github.com/openscholar/openscholar/issues/12452#issuecomment-555427431

We are passing the params in URL encoded now and were hoping you could add decoding to your JS code so that everything works as expected.  Please let me know if you have any questions.

@stamina84

**Comment by djbrooke** [comment]
Sure @rbran100 we'll take a look!

**Comment by pdurbin** [comment]
@rbran100 hi! While were making a pull request and thinking about OpenScholar, it might be a good time to finally add OpenScholar under our new-ish "integrations" page at http://guides.dataverse.org/en/4.18/admin/integrations.html

If you can suggest where OpenScholar "fits" on that page (a new category is fine) and can provide a blurb, it would be great!

Here are the places where OpenScholar is already mentioned in the User Guide:

- http://guides.dataverse.org/en/4.18/user/dataverse-management.html#adding-widgets-to-an-openscholar-website
- http://guides.dataverse.org/en/4.18/user/dataset-management.html#adding-widgets-to-an-openscholar-website

In addition to the guides, we're open to suggestions of how to describe OpenScholar support on our "features" page at https://dataverse.org/software-features

**Comment by Bilsi** [comment]
@pdurbin - we have documentation on how to integrate datavsere widgets in OpenScholar - https://help.theopenscholar.com/dataverse
It may be a good idea to make a new category called "OpenScholar" and add this content there.
Are you looking for content about OpenScholar that is unrelated to dataverse?
https://dataverse.org/software-features - It would be great if you could add us here. You can point it to theopenscholar.com or if you are looking for the features, then https://help.theopenscholar.com

**Comment by mheppler** [comment]
While updating the widgets code to work with the new OpenScholar D8 code, we should also make efforts to investigate issues with WordPress as outlined in _Widgets do not work with Wordpress #4357_.

**Comment by djbrooke** [comment]
- We will need some place to test on OS (is the new code available anywhere)

**Comment by rbran100** [comment]
@djbrooke we *could merge the POC code that we have so far but would rather test ourselves before merging in since there is much more going on on our side adding the widget code.  

Essentially the output is just a script tag so it should be pretty easy to test anywhere.  We really only have one publicly available test site for D8 live and would need to change it to point to your dev version to get a meaningful test going, is it possible to have the code staged somewhere and we could test compatibility with both D7 and D8 from our dev machines and then we can merge to our test servers for you to take a look if that works?

**Comment by djbrooke** [comment]
Thanks @rbran100. We'll talk a bit more in standup tomorrow and post more in this issue. This was just a quick note from sprint planning.

**Comment by pdurbin** [comment]
@rbran100 @stamina84 I'm just trying to wrap my head around this issue a bit.

Here's what I'm seeing on my production dataverse right now under "widgets":

## colon is not escaped

![Screen Shot 2019-12-17 at 9 48 15 AM](https://user-images.githubusercontent.com/21006/71005926-807a7d00-20b2-11ea-986b-f66e1cfb65b6.png)

At https://github.com/openscholar/openscholar/issues/12452#issuecomment-555427431 you talked about a change from...

`<script src="https://dataverse.harvard.edu//resources/js/widgets.js?persistentId=doi:10.7910/DVN/EBSVSE&amp;dvUrl=https://dataverse.harvard.edu/&amp;widget=citation&amp;heightPx=500"></script>`

... to ...

`<script src="https://dataverse.harvard.edu/resources/js/widgets.js?persistentId=doi%3A10.7910/DVN/EBSVSE&amp;dvUrl=https%3A//dataverse.harvard.edu/&amp;widget=citation&amp;heightPx=500"></script>`

What you're saying is that the colon (:) in "dvUrl" should be replaced with "%3A", right? And other escaping. Right? Here's a screenshot (I edited the colon manually):

## colon is escaped

![Screen Shot 2019-12-17 at 9 50 56 AM](https://user-images.githubusercontent.com/21006/71006167-e830c800-20b2-11ea-8fdd-bc8232a48dc8.png)

Are you saying you'd like to see this change on a test server? Does it need a valid cert? Any other requirements? Thanks!

By the way, I do agree with @mheppler that it would be good to make sure Dataverse works with Wordpress as well for #4357.

**Comment by stamina84** [comment]
@pdurbin Just created a test embed with Dataverse Dataset widget in Drupal 8 here: http://d8.theopenscholar.com/site01/people I hope it will help you to test something.

**Comment by pdurbin** [comment]
@stamina84 thanks, but how do I log in. With my site at https://scholar.harvard.edu/pdurbin I log in with HarvardKey like this:

![Screen Shot 2019-12-19 at 2 13 55 PM](https://user-images.githubusercontent.com/21006/71202311-f07f3380-2269-11ea-896c-aa7c9891f4e3.png)

At http://d8.theopenscholar.com/site01/people I'm not sure how to log in:

![Screen Shot 2019-12-19 at 2 14 05 PM](https://user-images.githubusercontent.com/21006/71202312-f117ca00-2269-11ea-9437-b5bf0732b4b7.png)

**Comment by pdurbin** [comment]
I'm able to log into http://d8.theopenscholar.com (thanks @rbran100 and @Bilsi) and created a page called "Data" but I can't figure out how to add any Dataverse widgets.

Here's how I add widgets in old/current OpenScholar:

Step 1: Edit Layout

![Screen Shot 2020-01-03 at 11 38 32 AM](https://user-images.githubusercontent.com/21006/71736015-03733580-2e1e-11ea-82a6-928acdc51df0.png)

Step 2: Add widget

![Screen Shot 2020-01-03 at 11 39 22 AM](https://user-images.githubusercontent.com/21006/71736016-03733580-2e1e-11ea-82c7-c1e77a3bc787.png)

With the new/d8 OpenScholar I can't figure out how to add any dataverse content. Here's how the "data" page I created at http://d8.theopenscholar.com/data looks.

![Screen Shot 2020-01-03 at 11 36 33 AM](https://user-images.githubusercontent.com/21006/71736013-03733580-2e1e-11ea-86c7-61d647b44252.png)

I can see Dataverse stuff under "Structure" like this:

![Screen Shot 2020-01-03 at 11 37 20 AM](https://user-images.githubusercontent.com/21006/71736014-03733580-2e1e-11ea-88e2-6c084a4bf0c4.png)

But I don't understand how to put the Dataverse widget on my new "Data" page.

For old/current OpenScholar, the process is well documented at https://help.theopenscholar.com/dataverse which looks like this:

![Screen Shot 2020-01-03 at 11 45 19 AM](https://user-images.githubusercontent.com/21006/71736245-957b3e00-2e1e-11ea-9e2e-1ecd4776b87b.png)

Are there docs for the new process? Or can we do a Zoom call so I can understand the new process?

**Comment by Bilsi** [comment]
@pdurbin - We haven't gotten to documentation yet. Click on place block
<img width="289" alt="Screen Shot 2020-01-03 at 11 47 55 AM" src="https://user-images.githubusercontent.com/11841641/71736417-fc005c00-2e1e-11ea-8880-4f907c792849.png">
and then "create a new widget"
<img width="318" alt="Screen Shot 2020-01-03 at 11 48 24 AM" src="https://user-images.githubusercontent.com/11841641/71736431-028ed380-2e1f-11ea-8da5-3c3491ba628c.png">
and you should see it

**Comment by pdurbin** [comment]
@Bilsi thank you!! I'm including some screenshots here because they might be useful for QA in the future.

Step 1: Click "Place Block"
![Screen Shot 2020-01-03 at 11 50 43 AM](https://user-images.githubusercontent.com/21006/71736574-6dd8a580-2e1f-11ea-9beb-b0245f8f3929.png)

Step 1: Click "Create New Widget"

![Screen Shot 2020-01-03 at 11 51 05 AM](https://user-images.githubusercontent.com/21006/71736575-6dd8a580-2e1f-11ea-96d7-197d15feb610.png)

**Comment by pdurbin** [comment]
@Bilsi sorry, I'm still really confused. Here's what happened on the d8 server when I tried to add a Dataverse widget:

![output9](https://user-images.githubusercontent.com/21006/71738208-f78a7200-2e23-11ea-9ea8-a5d0fa0b52af.gif)

(I don't do a lot of screen recordings. Here's what I did, for my reference: `ffmpeg -ss 00:00:04.000 -i d8.mov -pix_fmt rgb24 -filter:v "setpts=PTS/2" output9.gif`)

Here's a screenshot of console.log that might be more readable:

![Screen Shot 2020-01-03 at 12 26 12 PM](https://user-images.githubusercontent.com/21006/71738314-4932fc80-2e24-11ea-8acb-e65c13d78d41.png)

**Comment by stamina84** [comment]
Hi @pdurbin 
Its not allowed to create a block in "global" page, so I created a vsite (iqss) for you, please test it at following page: http://d8.theopenscholar.com/iqss/data-vsite
It should work now

**Comment by pdurbin** [comment]
@stamina84 much better! Thanks!

![Screen Shot 2020-01-03 at 12 58 27 PM](https://user-images.githubusercontent.com/21006/71739912-c06a8f80-2e28-11ea-9178-fb87e0858beb.png)

**Comment by pdurbin** [comment]
So far so good. I was able to add my "open-source-at-harvard" dataverse but I noticed that https://dataverse.harvard.edu seems to be assumed...

![Screen Shot 2020-01-03 at 1 00 07 PM](https://user-images.githubusercontent.com/21006/71740118-628a7780-2e29-11ea-8d09-b78004391b06.png)

... what about the other 52 installations of Dataverse? Can they integrate with OpenScholar?

![Screen Shot 2020-01-03 at 1 03 41 PM](https://user-images.githubusercontent.com/21006/71740145-8221a000-2e29-11ea-9274-03b25bd913ce.png)

**Comment by pdurbin** [comment]
> https://dataverse.harvard.edu seems to be assumed

In the OpenScholar code I found a variable called `dataverse_base_url`...

![Screen Shot 2020-01-03 at 1 29 51 PM](https://user-images.githubusercontent.com/21006/71741573-3d980380-2e2d-11ea-8615-8602b31435b8.png)

... but I couldn't find where to configure it in the OpenScholar UI. I'd like to be able to update `dataverse_base_url` to point at my test server, for example, assuming that's the way to point OpenScholar at a different installation of Dataverse.

**Comment by Bilsi** [comment]
> So far so good. I was able to add my "open-source-at-harvard" dataverse but I noticed that https://dataverse.harvard.edu seems to be assumed...
> 
> ![Screen Shot 2020-01-03 at 1 00 07 PM](https://user-images.githubusercontent.com/21006/71740118-628a7780-2e29-11ea-8d09-b78004391b06.png)
> 
> ... what about the other 52 installations of Dataverse? Can they integrate with OpenScholar?
> 
> ![Screen Shot 2020-01-03 at 1 03 41 PM](https://user-images.githubusercontent.com/21006/71740145-8221a000-2e29-11ea-9274-03b25bd913ce.png)

@pdurbin - all installations of DV can be integrated into the current version of OS. We expect it to work the same way in d8. The DV identifier is unique isn't it?

**Comment by pdurbin** [comment]
@Bilsi no, the "identifier" (or "alias") of a dataverse is not unique across installations of Dataverse. For example, here's my "open-source-at-harvard" dataverse in Harvard Dataverse:

![Screen Shot 2020-01-03 at 1 43 21 PM](https://user-images.githubusercontent.com/21006/71742330-5e615880-2e2f-11ea-9fbb-4139ee43818c.png)

And here's an "open-source-at-harvard" dataverse on a test server:
![Screen Shot 2020-01-03 at 1 44 06 PM](https://user-images.githubusercontent.com/21006/71742331-5e615880-2e2f-11ea-8dbc-da6e7f98e97f.png)

There could be a "cfa" dataverse in Harvard Dataverse that's for the Center for Astrophysics ( https://dataverse.harvard.edu/dataverse/cfa exists actually ) and a "cfa" dataverse in some other installation that's for Code for America (we could pretend https://dataverse.unc.edu/dataverse/cfa existed at UNC Dataverse).

I'd be happy to explain more in a Zoom call if that's easier. In short:

- I'm interested in understanding how the other 52 installations of Dataverse can configure OpenScholar to point at their installations of Dataverse. Perhaps this will help drive adoption for both OpenScholar and Dataverse.
- I'm interested in learning how to configure the "d8" installation of OpenScholar (using the `dataverse_base_url` variable?) to point to various installations of Dataverse for development purposes. That way, I can point the "d8" server to my "dev2" server, for example.

**Comment by stamina84** [comment]
@pdurbin I think I get it know, what you mean. So the main domain (`https://dataverse.harvard.edu/dataverse`) can be configure in D8 too, but there is no admin UI yet, we have to set it manually in D8 config per vsite (currently it is the same for all vsite)

More detailed we have the following separated config properties:
```
base_url: 'https://dataverse.harvard.edu/'
listing_base_url: 'https://dataverse.harvard.edu/dataverse/'
search_base_url: 'https://dataverse.harvard.edu/dataverse.xhtml'
```
`base_url` is used in Dataverse Dataset widget
`listing_base_url` is used in Dataverse Listing widget
and `search_base_url` in Dataverse Search Box widget

@Bilsi Do we need an admin UI to change these configs per vsite?
Is there any vsite, that has the following use case:

Place 2 Dataverse Dataset widgets and have following base urls:
- `https://dataverse.harvard.edu/dataverse/widget1identifier`
- `https://dataverse.harvard.edu/dataverse/widget2identifier`
So need to change the identifier per widget and not per vsite.

**Comment by pdurbin** [comment]
> can be configure in D8 too, but there is no admin UI yet, we have to set it manually in D8 config per vsite (currently it is the same for all vsite)

Great! I'm quite happy to configure the d8 server over and over as needed (to switch between dev Dataverse servers and prod Dataverse servers) if you explain how. I'm happy to ssh into the d8 server (you can get my public keys from https://github.com/pdurbin.keys ), edit files at the command line, run SQL statements, restart OpenScholar, etc.

This would also be good for other Dataverse installations to know how to do so that if one of the other 52 installations of Dataverse installs OpenScholar, they can reconfigure their OpenScholar installation to point at their Dataverse installation.

I hope this makes sense. You can also find me at https://chat.dataverse.org if it's easier to talk in real time. Or I can host a Zoom call. Or both. 😄

**Comment by stamina84** [comment]
@pdurbin I would not recommend to set these configs over SSH per vsite, I think it would be much better then to create an admin UI and set there. There is no simple drush command or simple SQL to override this in vsite context.
@Bilsi What do you think about that, shall I create a new issue for that and create an admin UI to able to set these urls?

@pdurbin BTW, what about the original problem, that mentioned the url decode in GET param?

**Comment by pdurbin** [comment]
@stamina84 I just created pull request #6486 and I'd love you to take a look.

@djbrooke as we discussed, you'll be figuring out some of the testing strategy. Thanks!

**Comment by djbrooke** [comment]
Thanks @pdurbin 

@stamina84 and @Bilsi, it would be much easier for us to test (and would be good for other installations) if the value for the dataverse installation could be easily configured. Thanks for thinking about this.

**Comment by stamina84** [comment]
@pdurbin We are ready with URL customization, and its deployed to D8 demo. You can edit all URLs at `/iqss/cp/settings/global-settings/dataverse_urls` admin url.

**Comment by pdurbin** [comment]
@stamina84 http://d8.theopenscholar.com/iqss/cp/settings/global-settings/dataverse_urls looks great! Thank you!! 🎉 Screenshot below.

![Screen Shot 2020-01-09 at 11 03 26 AM](https://user-images.githubusercontent.com/21006/72083384-bdabe680-32cf-11ea-803b-5a5923d54a73.png)

@djbrooke I see you took this issue out of the sprint. Do you want it back in?

**Comment by djbrooke** [comment]
@pdurbin yes, if it's unblocked let's pull it back in. 

@stamina84 thanks!

**Comment by pdurbin** [comment]
I just tested pull request #6486 against http://d8.theopenscholar.com/iqss/ and had a Zoom call with @Bilsi that I wrote about in https://github.com/IQSS/dataverse/pull/6486#issuecomment-573198995 . I put it back in code review and explained some expected bugs in OpenScholar (this is a development version) for QA.

I still think it would be nice to document this integration better. Perhaps this would be better as a separate pull request.

I didn't re-test the "Widgets do not work with Wordpress" issue at #4357 but @djbrooke and I did quickly scan through that issue. It looks like I tried and failed to get a Dataverse widget working in Wordpress in July 2018: https://github.com/IQSS/dataverse/issues/4357#issuecomment-400886893

As discussed in standup today, I did re title this issue a bit.

**Comment by pdurbin** [comment]
I went ahead and added some more to the Dataverse guides about OpenScholar in 4abf83c in the same pull request #6486.

On the integrations page I added a new category called "Embedding Data on Websites":

![Screen Shot 2020-01-10 at 4 24 01 PM](https://user-images.githubusercontent.com/21006/72187448-b0712380-33c5-11ea-9bbc-f409dc7884a3.png)
